UE/Application/SmsDb.cpp

```diff
@@ -24,4 +24,9 @@ const Sms& SmsDb::getMessage(gsl::index i)
     return messages.at(i);
 }
 
+void SmsDb::markedLastSmsAsFailed()
+{
+    messages.at(messages.size() - 1).state = SmsState::Failed;
+}
```

Comment:
In the specification it says ```markLastSmsSentAsFailed```. What's missing here is the ```Sent``` part. This method doesn't check whether the last SMS was in ```Sent``` state, thus it could potentially change state of wrong SMS that was received at the last moment. Also change "marked" to "mark" as it is confusing.

I would suggest using ```std::find_if``` with reverse iterator (```rbegin()```), that will return iterator to the first element it finds. Throw exception when it doesn't find anything as it is probably a logic error.

This function is very specific. I think that it is worth considering moving this logic to the user side.